File: java/array_search/InterpolationSearchProduction.java
import java.util.List;

public final class InterpolationSearchProduction {
    public record SearchResult(boolean found, int index, int insertionPoint) {
    }

    private InterpolationSearchProduction() {
    }

    public static SearchResult interpolationSearchFirst(List<Integer> values, int target) {
        if (values == null) {
            throw new IllegalArgumentException("values must not be null");
        }

        int length = values.size();
        int low = 0;
        int high = length - 1;
        int candidateLeft = 0;

        while (low <= high && values.get(low) <= target && target <= values.get(high)) {
            int lowValue = values.get(low);
            int highValue = values.get(high);

            if (lowValue == highValue) {
                boolean found = lowValue == target;
                return new SearchResult(found, found ? low : -1, low);
            }

            long numerator = ((long) target - lowValue) * (high - low);
            long denominator = (long) highValue - lowValue;
            int probe = low + (int) (numerator / denominator);
            probe = Math.max(low, Math.min(high, probe));
            int value = values.get(probe);

            if (value < target) {
                low = probe + 1;
                candidateLeft = low;
            } else {
                high = probe;

                if (value == target) {
                    break;
                }
            }
        }

        int searchRight = high >= candidateLeft ? high + 1 : length;
        int insertionPoint = lowerBound(values, target, candidateLeft, searchRight);
        boolean found = insertionPoint < length && values.get(insertionPoint) == target;

        return new SearchResult(found, found ? insertionPoint : -1, insertionPoint);
    }

    private static int lowerBound(List<Integer> values, int target, int left, int right) {
        while (left < right) {
            int mid = left + (right - left) / 2;

            if (values.get(mid) < target) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }

        return left;
    }
}

File: python/array_search/interpolation_search_production.py
from dataclasses import dataclass


@dataclass(frozen=True)
class InterpolationSearchResult:
    found: bool
    index: int
    insertion_point: int


def _lower_bound(values, target, left, right):
    while left < right:
        mid = left + (right - left) // 2

        if values[mid] < target:
            left = mid + 1
        else:
            right = mid

    return left


def interpolation_search_first(values, target):
    length = len(values)
    low = 0
    high = length - 1
    candidate_left = 0

    while low <= high and values[low] <= target <= values[high]:
        if values[low] == values[high]:
            found = values[low] == target
            return InterpolationSearchResult(found, low if found else -1, low)

        numerator = (target - values[low]) * (high - low)
        denominator = values[high] - values[low]
        probe = low + numerator // denominator
        probe = max(low, min(high, probe))
        value = values[probe]

        if value < target:
            low = probe + 1
            candidate_left = low
        else:
            high = probe

            if value == target:
                break

    insertion_point = _lower_bound(values, target, candidate_left, high + 1 if high >= candidate_left else length)
    found = insertion_point < length and values[insertion_point] == target

    return InterpolationSearchResult(
        found,
        insertion_point if found else -1,
        insertion_point
    )
